src/entities/removeLiquidity/removeLiquidityV3/index.ts

```diff
@@ -0,0 +1,228 @@
+import { createPublicClient, encodeFunctionData, http } from 'viem';
+
+import { balancerRouterAbi } from '@/abi';
+import { Token } from '@/entities/token';
+import { TokenAmount } from '@/entities/tokenAmount';
+import { PoolState } from '@/entities/types';
+import { getSortedTokens } from '@/entities/utils';
+import { Hex } from '@/types';
+import {
+    BALANCER_ROUTER,
+    CHAINS,
+    MAX_UINT112,
+    removeLiquiditySingleTokenExactInShouldHaveTokenOutIndexError,
+    removeLiquidityUnbalancedNotSupportedOnV3,
+} from '@/utils';
+
+import { getAmountsCall } from '../helper';
+import {
+    RemoveLiquidityBase,
+    RemoveLiquidityBaseCall,
+    RemoveLiquidityBaseQueryOutput,
+    RemoveLiquidityBuildOutput,
+    RemoveLiquidityInput,
+    RemoveLiquidityKind,
+} from '../types';
+
+export class RemoveLiquidityV3 implements RemoveLiquidityBase {
+    public async query(
+        input: RemoveLiquidityInput,
+        poolState: PoolState,
+    ): Promise<RemoveLiquidityBaseQueryOutput> {
+        const sortedTokens = getSortedTokens(poolState.tokens, input.chainId);
+        const bptToken = new Token(input.chainId, poolState.address, 18);
+
+        const client = createPublicClient({
+            transport: http(input.rpcUrl),
+            chain: CHAINS[input.chainId],
+        });
+
+        let bptIn: TokenAmount;
+        let amountsOut: TokenAmount[];
+        let tokenOutIndex: number | undefined;
+
+        switch (input.kind) {
+            case RemoveLiquidityKind.Unbalanced:
+                throw removeLiquidityUnbalancedNotSupportedOnV3;
+            case RemoveLiquidityKind.SingleTokenExactOut:
+                {
```

Comment:
Is it possible to make use of `getAmountsQuery` helper?

Reply:
I believe so 👍  I'll refactor over here.

Reply:
done ✅